packages/@react-spectrum/tag/src/Tag.tsx

```diff
@@ -10,82 +10,75 @@
-  const {
-    isDisabled: isGroupDisabled,
-    isRemovable: isGroupRemovable,
-    validationState: groupValidationState,
-    onRemove,
-    role
-  } =  useTagGroupProvider();
-
-  let removable = isGroupRemovable !== undefined ? isGroupRemovable : isRemovable;
-  let disabled = isGroupDisabled !== undefined ? isGroupDisabled : isDisabled;
-  let isInvalid = (validationState !== undefined ? validationState : groupValidationState) === 'invalid';
+  let {isFocusVisible, focusProps} = useFocusRing({within: true});
   let {clearButtonProps, labelProps, tagProps} = useTag({
     ...props,
-    isRemovable: removable,
-    isDisabled: disabled,
-    validationState: validationState !== undefined ? validationState : groupValidationState,
-    onRemove: props.onRemove || onRemove,
-    role
-  });
-  let {role: buttonRole, ...otherButtonProps} = clearButtonProps;
-  let icon = props.icon || (isInvalid && <Alert />);
+    isRemovable,
+    isDisabled,
+    item,
+    onRemove
+  }, state);
 
   return (
-    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
-      <div
-        {...styleProps}
-        {...mergeProps(tagProps, hoverProps)}
-        className={classNames(
-          styles,
-          'spectrum-Tags-item',
-          {
-            'is-disabled': disabled,
-            // 'is-selected': isSelected,
-            'spectrum-Tags-item--removable': removable,
-            'is-invalid': isInvalid,
-            'is-hovered': isHovered
-          },
-          styleProps.className
-        )}>
-        {icon && React.cloneElement(icon, {size: 'S', UNSAFE_className: classNames(styles, 'spectrum-Tags-itemIcon')})}
-        <span
-          {...labelProps}
-          className={classNames(styles, 'spectrum-Tags-itemLabel')}>
-          {props.children}
-        </span>
-        {removable &&
-          <span role={buttonRole}>
-            <ClearButton
-              tabIndex={tagProps.tabIndex}
-              focusClassName={classNames(styles, 'is-focused')}
-              UNSAFE_className={classNames(styles, 'spectrum-Tags-itemClearButton')}
-              {...otherButtonProps} />
-          </span>
+    <div
+      {...mergeProps(tagProps, hoverProps, focusProps, labelProps, otherProps)}
+      role={tagProps.role}
+      className={classNames(
+        styles,
+        'spectrum-Tags-item',
+        {
+          'is-disabled': isDisabled,
+          'focus-ring': isFocusVisible,
+          'is-focused': isFocusVisible,
+          'spectrum-Tags-item--removable': isRemovable,
+          'is-hovered': isHovered
         }
-      </div>
-    </FocusRing>
+      )}>
+      <SlotProvider
+        slots={{
+          icon: {UNSAFE_className: classNames(styles, 'spectrum-Tag-icon'), size: 'XS'},
+          text: {UNSAFE_className: classNames(styles, 'spectrum-Tag-content')}
+        }}>
+        {typeof props.children === 'string' ? <Text><span role="gridcell">{props.children}</span></Text> : props.children}
```

Comment:
When an icon gets rendered in addition to the text, the icon and text should both be wrapped within the element with `role="gridcell"`. And since the text content is provides the header for the row, we may want to use `role="rowheader"` instead of `role="gridcell"`.

Reply:
@tiffany-sy-lin Should be:

```html
<div role='row'> // wrapper that goes around all the tag parts like icon, text, etc
  <div role='rowheader'> // the container for the icon and text
  </div>
  <div role='gridcell'> // the container for the clear button
  </div>
</div>
```